### PR TITLE
feat: Share code with RIOT OS

### PR DESCRIPTION
**Pull request by chrysn**
# Description

This adds a test that takes the RIOT CoAP demo crate (which is also used in the `rust-gcoap` example of RIOT OS), and runs it on top of ariel-os-coap.

I think this is not good example material by the criteria of pedagogics and simplicity, so it's sitting in tests right now; I think that's a good staging area for something that can become an example at some point, but isn't quite there yet.

Note that this explicitly allows using git sources for those modules. Curiously they still live at gitlab.com/etonomy, where riot-wrappers was initially developed; those modules were not moved in with github.com/riot-os yet. They probably should be, but RIOT reviewers there didn't complain; whenever that module moves, our reference will follow.

## Change checklist

- [x] I have cleaned up my commit history and squashed fixup commits.
- [x] I have followed the [Coding Conventions](https://ariel-os.github.io/ariel-os/dev/docs/book/coding-conventions.html).
- [x] I have performed a self-review of my own code.
- [x] I have made corresponding changes to the documentation.

### CODE REVIEW

**Comment by kaspar030** [comment]
I hoped that the tests in `tests/` could be run automatic (think CI), given the right hardware is connected.

**Comment by chrysn** [comment]
There are two kinds of tests for which that doesn't quite fit:
* tests that are mainly build tests (particularly relevant for API guarantees) but can't be tested without human intervention (like, "does the display really show that?")
* tests that require host-side interaction (sending something over some network interface)

This one is a bit of both. Evolving it into fully-automated is certainly an option (with a CoAP client over loopback, and maybe hardware where two pins are shorted, or the pin is provided through a mutex) – but going through those lengths it should then rather be split up into smaller tests (to the point where the test on the RIOT components becomes moot here). Thus I'd suggest we treat this as a build test for now.

**Comment by chrysn** [comment]
Ping @kaspar030, now that I've started to write an FAQ entry that mentions this.

**Comment by chrysn** [comment]
Updated because main changed and now has changed w/rt how network is set up.

Pinging @kaspar030 for review, not only because it is pending, but also because the desire for a coap-blinky example came up again today.

This is still a test and not an example because its security is just not quite there yet, but as security stuff is just WIP, I'd rather leave it here and then do the simple move once #588 is in.

**Comment by chrysn** [comment]
The example has some potential for improvement around the GPIO semantics, but at that point we may want to coordinate with the blinky example on using switch-hal, which itself [needs some love](https://github.com/rubberduck203/switch-hal/issues/13).

File: tests/coap-blinky/laze.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+apps:
+  - name: coap-blinky
+    env:
+      global:
+        CARGO_ENV:
+          - CONFIG_ISR_STACKSIZE=16384
+    selects:
+      - network
+      - random
```
[the remark this answers is not in the document]
Comment (chrysn) [reply tests/coap-blinky/laze.yml:9]: CoAP uses random internally. Does this indicate that we this dependency should be threaded through laze?

File: tests/coap-blinky/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+authors = ["Christian Amsüss <chrysn@fsfe.org>"]
+license.workspace = true
+edition.workspace = true
+publish = false
+
+[lints]
+workspace = true
+
+[dependencies]
+embassy-sync = { workspace = true }
```
[the remark this answers is not in the document]
Comment (chrysn) [reply tests/coap-blinky/Cargo.toml:12]: Indeed not, thanks, removing.

File: tests/coap-blinky/laze.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (chrysn) [reply tests/coap-blinky/laze.yml:9]: There may be some corner cases in which they can be made to not depend on it, but it's very hard to carve those out, so yes, generally those do depend on random.

File: tests/coap-blinky/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+] }
+ariel-os-boards = { path = "../../src/ariel-os-boards" }
+coap-handler-implementations = "0.5.0"
+
+riot-coap-handler-demos = { git = "https://gitlab.com/etonomy/riot-module-examples", rev = "09fa2d45e92ca4e46da7f18af3db3d1bcec238d3", default-features = false, features = [
+  "gpio",
+] }
+
+# Just for the network config
+embassy-net = { workspace = true, features = ["udp"] }
```
[the remark this answers is not in the document]
Comment (chrysn) [reply tests/coap-blinky/Cargo.toml]: Adjusted in both now.

File: tests/coap-blinky/laze.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (chrysn) [reply tests/coap-blinky/laze.yml:9]: Given that this is used that way not only in the other CoAP test but also in the examples, plus needs extra input, I'd rather fix that properly in all 4 occurrences after this is merged.

Then, we can also think of whether coap should pull in network on its own, and whether it should pull in udp too, and what that means for doing a manual network setup in the examples.

**Comment by chrysn** [comment]
All that I think make sense to do here are addressed now; can I rebase?

File: tests/coap-blinky/laze.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (chrysn) [reply tests/coap-blinky/laze.yml:9]: Tracking this in https://github.com/ariel-os/ariel-os/issues/674

**Comment by chrysn** [comment]
Rebasing hit some cases of "things work differently now"; CI looks good so far.

**Review by kaspar030 — APPROVED** [review]
fine
